[JSC] Do not resolve rope when constructing Function

Constellation opened the pull request:
#### 05cdfcbb72aa9ab91d3d62a2ac8fc8a2d64ee694
<pre>
[JSC] Do not resolve rope when constructing Function
<a href="https://bugs.webkit.org/show_bug.cgi?id=294878">https://bugs.webkit.org/show_bug.cgi?id=294878</a>
<a href="https://rdar.apple.com/154145026">rdar://154145026</a>

Reviewed by Yijia Huang.

Let&apos;s avoid resolving string for input for Function constructor.
It is not so super efficient if we resolve them. We also avoid creating
SourceCode so that we do not need to allocate StringSourceProvider.

* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::stringifyFunction):
(JSC::constructFunction):
(JSC::constructFunctionSkippingEvalEnabledCheck):
* Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
(JSC::FunctionExecutable::fromGlobalCode):
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::tryGetCachedFunctionExecutableForFunctionConstructor):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSString.h:
(WTF::StringTypeAdapter&lt;JSC::JSString::StringTypeAdapter):
(WTF::StringTypeAdapter&lt;JSC::JSString::length const):
(WTF::StringTypeAdapter&lt;JSC::JSString::is8Bit const):
(WTF::StringTypeAdapter&lt;JSC::JSString::writeTo const):
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::JSString::resolveToBuffer):

Canonical link: <a href="https://commits.webkit.org/296661@main">https://commits.webkit.org/296661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fd257f379f2c481e6b596ed87e1541c919e631e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109101 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28761 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19186 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114312 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59400 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29444 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37328 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82918 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112049 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23423 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98266 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63363 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22824 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16408 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58995 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101630 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92794 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16451 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117430 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107659 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36149 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26721 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91933 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36520 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94530 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91739 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36647 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14400 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32007 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17623 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36046 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41551 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131930 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35739 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35759 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37425 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/34349b035ff1b60b5bf9286aef269ee72876a166. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47096

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9fd257f379f2c481e6b596ed87e1541c919e631e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47096

Constellation committed:
[JSC] Do not resolve rope when constructing Function
https://bugs.webkit.org/show_bug.cgi?id=294878
rdar://154145026

Reviewed by Yijia Huang.

Let's avoid resolving string for input for Function constructor.
It is not so super efficient if we resolve them. We also avoid creating
SourceCode so that we do not need to allocate StringSourceProvider.

* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::stringifyFunction):
(JSC::constructFunction):
(JSC::constructFunctionSkippingEvalEnabledCheck):
* Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
(JSC::FunctionExecutable::fromGlobalCode):
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::tryGetCachedFunctionExecutableForFunctionConstructor):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSString.h:
(WTF::StringTypeAdapter<JSC::JSString::StringTypeAdapter):
(WTF::StringTypeAdapter<JSC::JSString::length const):
(WTF::StringTypeAdapter<JSC::JSString::is8Bit const):
(WTF::StringTypeAdapter<JSC::JSString::writeTo const):
* Source/JavaScriptCore/runtime/JSStringInlines.h:
(JSC::JSString::resolveToBuffer):

Canonical link: https://commits.webkit.org/296661@main

webkit-commit-queue commented:
Committed 296661@main (05cdfcbb72aa): <https://commits.webkit.org/296661@main>

Reviewed commits have been landed. Closing PR #47096 and removing active labels.
